### PR TITLE
[AMDGPU] Fix legalization of frem(-0.0, y)

### PR DESCRIPTION
**Pull request by jayfoad**
We legalized frem(x, y) -> x - y * trunc(x / y). When x is -0.0 this
evaluates to +0.0 but the result should be -0.0.

Fix this by legalizing to copysign(x - y * trunc(x / y), x).

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/GlobalISel/frem.ll [hunk 1]
```diff
@@ -29,7 +29,10 @@ define amdgpu_kernel void @frem_f16(ptr addrspace(1) %out, ptr addrspace(1) %in1
 ; CI-NEXT:    s_mov_b32 s7, 0xf000
 ; CI-NEXT:    v_div_fixup_f32 v2, v2, v1, v0
 ; CI-NEXT:    v_trunc_f32_e32 v2, v2
-; CI-NEXT:    v_fma_f32 v0, -v2, v1, v0
+; CI-NEXT:    v_fma_f32 v1, -v2, v1, v0
+; CI-NEXT:    v_and_b32_e32 v1, 0x7fffffff, v1
+; CI-NEXT:    v_and_b32_e32 v0, 0x80000000, v0
+; CI-NEXT:    v_or_b32_e32 v0, v1, v0
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/frem.ll:35]: Note that the fix is applied here in `frem_f16` and in `unsafe_frem_f16` but not in `fast_frem_f16`.

**Comment by jayfoad** [comment]
@b-sumner FYI.

**Comment by jayfoad** [comment]
It has been pointed out to me that this also changes the sign of the result when the division `x / y` overflows.

Previously: frem(huge, tiny) -> infinity with opposite sign from x
Now: frem(huge, tiny) -> infinity with same sign as x

I'm not sure if this is a good thing or a bad thing.

**Comment by arsenm** [comment]
I think the frem lowering is just broken in general for large values

**Review by arsenm — APPROVED** [review]
Pretty sure frem is just generally broken, and this at least fixes one way it's broken. We never wired up OpenCL to use frem for example, so I didn't know anything was actually using this.

**Comment by jayfoad** [comment]
The motivation for this has gone away. I wanted it for Vulkan/SPIR-V but it turns out that SPIR-V does not specify the sign of a zero result from OpFRem.

**Comment by jayfoad** [comment]
Actually there is still interest in this, either for its own sake (to follow the LLVM definition in LangRef) or because the SPIR-V definition may change to match the LLVM definition.

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 2]
```diff
@@ -2341,6 +2341,14 @@ bool AMDGPULegalizerInfo::legalizeFceil(
...
 }
 
+static bool allowNoSignedZeros(const MachineFunction &MF, unsigned Flags) {
+  return (Flags & MachineInstr::FmNsz) ||
+         MF.getTarget().Options.NoSignedZerosFPMath;
+}
+
+// Legalize frem(x, y) -> copysign(x - y * trunc(x / y), x)
+// The copysign is only required to get the correct result -0.0 when x is -0.0
+// (and y is non-zero). With NSZ it can be dropped.
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:2351]: Could also drop it if isKnownNeverZero(x)?

**Comment by arsenm** [comment]
Should probably do this if it follows the LangRef. But again, this lowering is wrong to begin with

**Pull request closed without merging**